Source/VCGeneration/LoopExtractor.cs

```diff
@@ -0,0 +1,621 @@
+      blockMap.Remove(header);
+
+      Contract.Assert(fullMap[impl.Name][header.Label] == header);
+      fullMap[impl.Name][header.Label] = origHeader;
+
+      foreach (Block block in blockMap.Keys)
+      {
+        // Don't add dummy blocks to the map
+        if (dummyBlocks.Contains(blockMap[block].Label))
+        {
+          continue;
+        }
+
+        // Following two statements are for nested loops: compose map
+        if (!fullMap[impl.Name].ContainsKey(block.Label))
+        {
+          continue;
+        }
+
+        var target = fullMap[impl.Name][block.Label];
+
+        AddToFullMap(fullMap, loopProc.Name, blockMap[block].Label, target);
+      }
+
+      fullMap[impl.Name].Remove(header.Label);
+      fullMap[impl.Name][lastIterBlockName] = origHeader;
+    }
+  }
+
+  private static void addUniqueCallAttr(int val, CallCmd cmd)
+  {
+    var a = new List<object>();
+    a.Add(new LiteralExpr(Token.NoToken, Microsoft.BaseTypes.BigNum.FromInt(val)));
+
+    cmd.Attributes = new QKeyValue(Token.NoToken, "si_unique_call", a, cmd.Attributes);
+  }
+
+  private static void AddToFullMap(Dictionary<string, Dictionary<string, Block>> fullMap, string procName, string blockName,
+    Block block)
+  {
+    if (!fullMap.ContainsKey(procName))
+    {
+      fullMap[procName] = new Dictionary<string, Block>();
+    }
+
+    fullMap[procName][blockName] = block;
+  }
+
+  private static HashSet<Block> GetBlocksInAllNaturalLoops(CoreOptions options, Block header, Graph<Block /*!*/> /*!*/ g)
+  {
+    Contract.Assert(options.DeterministicExtractLoops,
+      "Can only be called with /deterministicExtractLoops option");
+    var allBlocksInNaturalLoops = new HashSet<Block>();
+    foreach (Block /*!*/ source in g.BackEdgeNodes(header))
+    {
+      Contract.Assert(source != null);
+      g.NaturalLoops(header, source).Iter(b => allBlocksInNaturalLoops.Add(b));
+    }
+
+    return allBlocksInNaturalLoops;
+  }
+
+
+  /// <summary>
+  /// Finds blocks that break out of a loop in NaturalLoops(header, backEdgeNode)
+  /// </summary>
+  /// <param name="header"></param>
+  /// <param name="backEdgeNode"></param>
+  /// <returns></returns>
+  private static HashSet<Block> GetBreakBlocksOfLoop(CoreOptions options, Block header, Block backEdgeNode, Graph<Block /*!*/> /*!*/ g)
+  {
+    Contract.Assert(options.DeterministicExtractLoops,
+      "Can only be called with /deterministicExtractLoops option");
+    var immSuccBlks = new HashSet<Block>();
+    var loopBlocks = g.NaturalLoops(header, backEdgeNode);
+    foreach (Block /*!*/ block in loopBlocks)
+    {
+      Contract.Assert(block != null);
+      var auxCmd = block.TransferCmd as GotoCmd;
+      if (auxCmd == null)
+      {
+        continue;
+      }
+
+      foreach (var bl in auxCmd.labelTargets)
+      {
+        if (loopBlocks.Contains(bl))
+        {
+          continue;
+        }
+
+        immSuccBlks.Add(bl);
+      }
+    }
+
+    return immSuccBlks;
+  }
+}
```

Comment:
Moving things like this makes reviewing and tracing bugs harder.  It may still be worth it, but the usual git facilities for inspecting code break badly in the face of this.

Reply:
Originally it was sufficient to review the refactor commits separately from the other commits, but I've since added a merge commit in which several conflicts were resolved. I'm not sure what the best workflow is here. Maybe I should have done a rebase instead then you could have still reviewed the individual commits.

Reply:
Github having trouble is fine, because I can review locally.  What I'm more concerned about is `git-log -L`, which is extremely useful for understanding the history of a function, and also completely broken by renames / moves.